Change File::try_lock() and try_lock_shared() to return io::Result<()>

cberner opened the pull request:
This PR changes the signatures of `File::try_lock` and `File::try_lock_shared` as requested by @rust-lang/libs-api in this discussion: https://github.com/rust-lang/rust/issues/130994#issuecomment-2855657455

These methods are unstable under the "file_lock" feature. The related tracking issue is https://github.com/rust-lang/rust/issues/130994

cberner committed:
Change File::try_lock() and try_lock_shared() to return io::Result<()>

Amanieu commented on library/std/src/sys/fs/windows.rs:
```diff
@@ -444,9 +443,9 @@ impl File {
                 if err.raw_os_error() == Some(c::ERROR_IO_PENDING as i32)
                     || err.raw_os_error() == Some(c::ERROR_LOCK_VIOLATION as i32) =>
```
@ChrisDenton Would it make sense to have these directly convert to `ErrorKind::WouldBlock`?

ChrisDenton replied:
For `ERROR_LOCK_VIOLATION` that would make sense.

For `ERROR_IO_PENDING` it would be very wrong. And in fact looking at the code here I'm pretty sure it's currently unsound, assuming `ERROR_IO_PENDING` were ever returned (which it shouldn't be with `LOCKFILE_FAIL_IMMEDIATELY`). The function *must not* return while there is a pending I/O operation. It must wait for it to complete because we've passed in a reference to the stack.

cberner replied:
@ChrisDenton ah, thanks for flagging that. I have very little experience with the Windows platform, so I appreciate the help!

The  [LockFileEx documentation](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-lockfileex#remarks) says:
>The LockFileEx function operates asynchronously if the file handle was opened for asynchronous I/O, unless the LOCKFILE_FAIL_IMMEDIATELY flag is specified. If an exclusive lock is requested for a range of a file that already has a shared or exclusive lock, the function returns the error ERROR_IO_PENDING.


and I interpreted that to mean that when `LOCKFILE_FAIL_IMMEDIATELY` is specified, then waiting on the event is unnecessary. But am I misreading that, and we need the same event waiting code that's in `acquire_lock()` in the `try_lock()`/`try_lock_shared()` functions?

ChrisDenton replied:
Yes, with `LOCKFILE_FAIL_IMMEDIATELY` waiting should not be necessary so checking for `ERROR_IO_PENDING` is also unnecessary. The code can just check for `ERROR_LOCK_VIOLATION `.

cberner replied:
Ah, got it. Thanks! Yes, I added a test for this case and it seems to pass just fine with `ERROR_IO_PENDING` removed.

And I added a separate commit that changes `decode_error_kind()` to map `ERROR_LOCK_VIOLATION` to `ErrorKind::WouldBlock` along with that test case. I assume that's the right place to do the conversion that you and @Amanieu were talking about?

joshtriplett commented:
This looks good to me, pending the question about Windows.

cberner committed:
Remove unnecessary handling of ERROR_IO_PENDING

try_lock() and try_lock_shared() do not need to handle these per the discussion in https://github.com/rust-lang/rust/pull/140718#discussion_r2076678485

cberner commented:
I think the question about Windows has been resolved now.

r? @joshtriplett

Amanieu commented on library/std/src/sys/pal/windows/mod.rs:
```diff
@@ -114,6 +114,7 @@ pub fn decode_error_kind(errno: i32) -> ErrorKind {
         c::ERROR_TOO_MANY_LINKS => return TooManyLinks,
         c::ERROR_FILENAME_EXCED_RANGE => return InvalidFilename,
         c::ERROR_CANT_RESOLVE_FILENAME => return FilesystemLoop,
+        c::ERROR_LOCK_VIOLATION => return WouldBlock,
```
On second thought, this is likely to be incorrect: this error can be returned from other operations such as opening a file that is write-locked. It would be better to revert to the old behavior of explicitly returning `ErrorKind::WouldBlock` only for `try_lock`.

cberner replied:
Ok, I reverted back to the old behavior. I see there's also some discussion in the [tracking issue](https://github.com/rust-lang/rust/issues/130994#issuecomment-2895342658) of also reverting back to the API with `TryLockError` and adding a `From` impl.

Amanieu commented:
Closing since the libs-api team decided to stay with the previous API in the end.
